torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1806,84 +1834,148 @@ struct to_ir {
+
+    RefinementSet refinement;
+
+    TypePtr unified_true = nullptr;
+    TypePtr unified_false = nullptr;
+
+    std::vector<TypePtr> isinstance_types;
+    std::vector<TypePtr> not_isinstance_types;
+
+    std::vector<Refinement> true_refinements;
+    std::vector<Refinement> false_refinements;
+
+    bool all_lhs_subtype_some_rhs = true;
+
+    // We can discard any rhs types that we know statically would be
+    // impossible. For example, if we had:
+    //
+    //    def fn(x: Optional[str]):
+    //        if isinstance(x, (List[str], str, int)):
+    //            ...
+    //
+    // then `x` would be `str` in the true branch and `None` in the
+    // false branch, not `(List[str], str, int)` in the true branch
+    // and `None` in the false branch
+    for (TypePtr lhs_type : lhs_types) {
+      if (lhs_type == AnyType::get()) {
+        isinstance_types.insert(
+            isinstance_types.end(), rhs_types.begin(), rhs_types.end());
+        not_isinstance_types.push_back(AnyType::get());
+        // Edge case: we can still say that all lhs types subtype some
+        // rhs type if `lhs` is `Any` and `rhs` is `Any`
+        if (isinstance_types.size() != 1 ||
+            isinstance_types[0] != AnyType::get()) {
+          all_lhs_subtype_some_rhs = false;
         }
-        return false;
+        break;
       }
-      bool maybeOfKind(TypeKind kind, const TypePtr& actual_type) {
-        if (actual_type->kind() == AnyType::Kind) {
-          return true;
-        }
-        if (const auto optional_type = actual_type->cast<OptionalType>()) {
-          return optional_type->getElementType()->kind() == kind;
+
+      auto get_smaller_type = [&](TypePtr t1, TypePtr t2) -> TypePtr {
+        if (!t1 && !t2) {
+          return nullptr;
+        } else if (!t1 || !t2) {
+          return t1 ? t1 : t2;
+        } else if (t1->isSubtypeOf(t2)) {
+          return t1;
+        } else if (t2->isSubtypeOf(t1)) {
+          return t2;
+        } else {
+          return nullptr;
         }
-        if (const auto union_type = actual_type->cast<UnionType>()) {
-          return std::any_of(
-              union_type->containedTypes().begin(),
-              union_type->containedTypes().end(),
-              [&](const TypePtr contained) {
-                return contained->kind() == kind;
-              });
+      };
+
+      TypePtr found_refinement = nullptr;
+      for (TypePtr rhs_type : rhs_types) {
+        TypePtr maybe_smaller_type = get_smaller_type(lhs_type, rhs_type);
+        if (!maybe_smaller_type) {
+          continue;
+        } else if (*maybe_smaller_type == *lhs_type) {
+          // Cover the case that we have something like
+          // lhs = `List[str]` and rhs = `list`
+          found_refinement = lhs_type;
+        } else if (*maybe_smaller_type == *rhs_type) {
+          // We want the narrowest possible type
+          found_refinement = get_smaller_type(found_refinement, rhs_type);
```

Comment:
So, I think this probably mostly works because of our limited type hierarchies, especially when you `standardizeVectorForUnion` before the call. 

However, i think the logic could be adjusted a bit. If you have multiple different rhs type which subtype a lhs type, the different rhs types do not necessarily have a type relation to each other. 
          t1
      /    |    \
t2       t3      t4

t2 <: t1
t3 <: t1
t4 <: t1

If we are doing isinstance(t1, (t2, t4)), when we process `t4` the refinement should be Union[t2, t4] instead of just t4. Right now `get_smaller_type(found_refinement, rhs_type)` will produce nullptr. 

I think we want to either initialize found_refinement if it hasn't be set yet, or set found_refinement to the union of  the existing type and rhs_type here. 
```
found_refinement = found_refinement ? unifyTypes(found_refinement, rhs_type) : rhs_type;
```